src/coreclr/jit/lower.cpp

```diff
@@ -2783,6 +2783,36 @@ GenTree* Lowering::LowerCompare(GenTree* cmp)
     }
 #endif
 
+    // Optimizes (X & 1) == 1 to (X & 1)
```

Comment:
This should live as a special case in `OptimizeConstCompare`:

https://github.com/dotnet/runtime/blob/9dd21e4cc95beab491d76c0a68dc466d9a2b0e11/src/coreclr/jit/lower.cpp#L2592-L2605

Currently, the checks for `EQ`/`NE` is missing. Additionally, transforming `(X & 1) != 1` to `x & 1` is not correct, I suspect that's the cause of test failures.

Reply:
What is wrong with the transforming? Did I implement the BlockReplacing wrong?
The `OptimizeConstCompare` method is guarded by `!comp->opts.MinOpts`
https://github.com/dotnet/runtime/blob/9dd21e4cc95beab491d76c0a68dc466d9a2b0e11/src/coreclr/jit/lower.cpp#L2816-L2818
Could you tell me what the opts are?

Reply:
> What is wrong with the transforming? Did I implement the BlockReplacing wrong?

No, the transformation itself looks correct to me (well, there is one issue - `GTF_RELOP_JMP_USED` is not guaranteed to be maintained by this phase, so it'll have to be replaced by a check of the user's oper). It is just happening under conditions that are too lax. `LowerCompare` handles all sorts of relops, so this code will also optimize, e. g. `(x & 1) > 1` to `x & 1`, which is not correct.

>  Could you tell me what the opts are?

There are a lot of options; `opts.MinOpts` in particular means that the compiler is not to perform (almost) any optimizations. You can see it checked in `opts.OptimizationEnable`. It will be set e. g. for Tier0 code, or for methods that are explicitly marked with `MethodImplOptions.NoOptimization`.

Reply:
> No, the transformation itself looks correct to me (well, there is one issue - `GTF_RELOP_JMP_USED` is not guaranteed to be maintained by this phase, so it'll have to be replaced by a check of the user's oper). It is just happening under conditions that are too lax. `LowerCompare` handles all sorts of relops, so this code will also optimize, e. g. `(x & 1) > 1` to `x & 1`, which is not correct.
>
I'm going to change `GTF_RELOP_JMP_USED` to
`cmpUse.User()->OperIs(GT_RETURN, GT_STOREIND, GT_STORE_LCL_VAR, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_LCL_FLD)`.
Is it correct? Do I understand you right? If yes then I'm concerned that I miss some operations or check wrong ones.

Reply:
> `cmpUse.User()->OperIs(GT_RETURN, GT_STOREIND, GT_STORE_LCL_VAR, GT_STORE_BLK, GT_STORE_OBJ, GT_STORE_LCL_FLD)`

No need for that; we only need to know if the relop is a jump, i. e. `User()->OperIs(GT_JTRUE)`.

Reply:
I saw this snippet across the lower.cpp and tried to use it, but `User()->OperIs(GT_JTRUE)` gives me false, because in my case `User()->OperGet` returns `GT_RETURN`. I also can't see the oper in the dump.
<img width="981" alt="image" src="https://user-images.githubusercontent.com/18483133/149145146-f1a8f798-d78c-4fc0-a8b7-da4dc440c040.png">
 What am I missing?

Reply:
Hmm, the idea would be that the check is actually `!User()->OperIs(GT_JTRUE)` (since it was `!(gtFlags & GTF_RELOP_JMP_USED)`).

Reply:
True, my bad! I missed it, thank you a lot!!!